panther_lights/src/controller_node.py

```diff
@@ -323,99 +324,124 @@ def _get_image_animation_description(self, animation: LEDImageAnimation) -> dict
+            animation = PantherAnimation()
 
-                    if not 'type' in anim_desc:
-                        raise KeyError('Missing \'type\' in animation description')
+            for panel in anim['animation']:
+                anim_desc = anim['animation'][panel]
 
-                    try:
-                        BASIC_ANIMATIONS[anim_desc['type']]
-                    except KeyError as err:
-                        raise KeyError(f'Undefined animation type: {err}')
+                if not 'type' in anim_desc:
+                    raise KeyError('Missing \'type\' in animation description')
 
-                    if panel == 'both':
-                        animation.front = BASIC_ANIMATIONS[anim_desc['type']](
-                            anim_desc, self._num_led, self._controller_frequency
-                        )
-                        animation.rear = BASIC_ANIMATIONS[anim_desc['type']](
-                            anim_desc, self._num_led, self._controller_frequency
-                        )
-                        break
-                    elif panel == 'front':
-                        animation.front = BASIC_ANIMATIONS[anim_desc['type']](
-                            anim_desc, self._num_led, self._controller_frequency
-                        )
-                    elif panel == 'rear':
-                        animation.rear = BASIC_ANIMATIONS[anim_desc['type']](
-                            anim_desc, self._num_led, self._controller_frequency
-                        )
-                    else:
-                        raise KeyError(f'Invalid panel type: {panel}')
+                try:
+                    BASIC_ANIMATIONS[anim_desc['type']]
+                except KeyError as err:
+                    raise KeyError(f'Undefined animation type: {err}')
 
-                if not hasattr(animation, 'front') or not hasattr(animation, 'rear'):
-                    raise KeyError(
-                        'Missing \'both\' or \'front\'/\'rear\' in animation description'
+                if panel == 'both':
+                    animation.front = BASIC_ANIMATIONS[anim_desc['type']](
+                        anim_desc, self._num_led, self._controller_frequency
                     )
+                    animation.rear = BASIC_ANIMATIONS[anim_desc['type']](
+                        anim_desc, self._num_led, self._controller_frequency
+                    )
+                    break
+                elif panel == 'front':
+                    animation.front = BASIC_ANIMATIONS[anim_desc['type']](
+                        anim_desc, self._num_led, self._controller_frequency
+                    )
+                elif panel == 'rear':
+                    animation.rear = BASIC_ANIMATIONS[anim_desc['type']](
+                        anim_desc, self._num_led, self._controller_frequency
+                    )
+                else:
+                    raise KeyError(f'Invalid panel type: {panel}')
```

Comment:
The code will ignore anything as long as there exists key `both`, but if there is `front` and `rear` it will proceed to check invalid keys. It could first check
``` python
if panel not in ('both', 'front', 'rear'):
    raise KeyError(f'Invalid panel type: {panel}')
```
This will solve the problem

Reply:
NVM. It's ok